### ISSUE TITLE
#[error]Could not validate drive id.

### ISSUE THREAD

**Issue by shiva-gattu** [halliba/az-pipelines-2-sharepoint#8]
I'm facing and error when i try to move files from Azure repos to sharepoint using this extension. I get an error as could not validate drive Id. not sure where I'm missing
- task: az-pipelines-2-sharepoint@0
  inputs:
    tenantId: X
    clientId: Y
    clientSecret: Z
    driveId: 'https://learn1212.sharepoint.com/sites/testprojectsite/Shared%20Documents'
    sourceFolder: 
    contents: 'build-files/**/*.pdf'
    conflictBehaviour: 'fail'
    cleanTargetFolder: false
    flattenFolders: false
    failOnEmptySource: false

**Comment by MHebes** [comment]
I combined https://learn.microsoft.com/en-us/azure/active-directory/develop/tutorial-v2-nodejs-console with your code (primarily the `drive-utils.ts`), and it seems to work great.

Specifically, I used the `@azure/msal-node` package to get a token instead of the raw ``fetch(`https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/token`, ...)`` that this library currently uses.

[This comment on an issue for onedrive's api](https://github.com/OneDrive/onedrive-api-docs/issues/785#issuecomment-420274451) suggests that you may be missing the `resource` query param in [getAccessTokenAsync](https://github.com/halliba/az-pipelines-2-sharepoint/blob/main/buildAndReleaseTask/src/uploader.ts#L50), but I'm not sure because your `grant_type` is different from that person's.

**Comment by MHebes** [comment]
Nope, that wasn't it. In my test script I switched back from `msal` to your `getAccessTokenAsync` without modifications and it works great.

**Comment by MHebes** [comment]
I was using the wrong pipeline variable for my client secret lol ignore all previous posts. Learned a lot about graph APIs.

**Comment by rdinizz** [comment]
I'm having a similar issue:

`##[error]{"statusCode":401,"code":"InvalidAuthenticationToken","requestId":"xyz","date":"2024Z","body":"{\"code\":\"InvalidAuthenticationToken\",\"message\":\"IDX14100: JWT is not well formed, there are no dots (.).\\nThe token needs to be in JWS or JWE Compact Serialization Format. (JWS): 'EncodedHeader.EndcodedPayload.EncodedSignature'. (JWE): 'EncodedProtectedHeader.EncodedEncryptedKey.EncodedInitializationVector.EncodedCiphertext.EncodedAuthenticationTag'.\",\"innerError\":{\"date\":\"2024\",\"request-id\":\"xyz\",\"client-request-id\":\"zyu\"}}"}`

I set the client secret in Library variable group as a secret, any idea of what i might be doing wrong?

**Comment by VassilisM** [comment]
Hi. Having the same issue.
Granted, not sure if the tenantID, clientID and secret are correct, there are so many in there.
I am using as tenantID the whole organization and got the clientID and secret from the Sharepoint application.
But regarding this "token", no idea where to get it from or where to use it.

**Comment by halliba** [comment]
Did you create an App Registration in Entra ID / Azure AD?
The token mentioned in the error message is nothing you can configure. It is acquired dynamically form Entra ID with the Client ID and Client Secret that you've configured.
It looks like the extension receives some invalid response from Entra ID and passes it on to Microsoft Graph, so the issue is probably in the App Registration or Permissions. I will try to replicate the problem....

**Comment by VassilisM** [comment]
I did not. I took the clientID and secret (well, IT generated a new one) from the registered service connection of the project.
If I were to create (I do not know if I have the rights yet), what would I do that for? As a sharepoint "proxy"?

**Comment by halliba** [comment]
Hi, yes it should be possible to reuse any other service principal, but I have not tested it. Also I'm not 100% sure, what you mean by "service principal of the project".
In any case, you have to make sure, that the app registration of the service principal has the `Files.ReadWrite.All` permission set, or the `Sites.Selected` with appropriate permissions on SharePoint. For the latter, I've heard from multiple users, that they had problems setting it up, so I have to verfiy that.. You might want to hint your IT to this repo's readme, they should be able to configure the app / service principal accordingly

**Comment by halliba** [comment]
Hi, to find the site-id required for the grant command, you can use a GET request with this format `/sites/{hostname}:/{server-relative-path}` in Graph Explorer. `hostname` would be `contoso.sharepoint.com` and `server-relative-path` would be `sites/my-site`.
I know, that there are problems with nested sites and the discovery of those via Graph Explorer - so that might be the problem. 
In General Microsoft Graph works best, if you can address the resource directly, not via the search capabilities.

See also these links:
https://stackoverflow.com/questions/45406451/how-can-i-get-the-siteid-of-the-current-site-with-microsoft-graph-api
https://learn.microsoft.com/en-us/graph/api/site-get?view=graph-rest-1.0&tabs=http#access-a-site-by-server-relative-url

**Comment by rdinizz** [comment]
Thank you @halliba i forgot to give an update but basically i found 2 ways of giving the permissions to an ad app:

It can be done either via Graph API
```
POST <-  
URL: https://graph.microsoft.com/v1.0/sites/tenant.sharepoint.com,siteId,webId/permissions
BODY:
{
    "roles": [
        "write"
    ],
    "grantedToIdentities": [
        {
            "application": {
                "id": "clientId",
                "displayName": "App Name" 
            }
        }
    ]
}
```
The siteId needed is in this format tenant.sharepoint.com,siteId,webId
I was not able to get the siteId via graph queries, but i managed to get it adding _api/site/id and _api/web/id at the end of the site url, example https://contoso.sharepoint.com/sites/some-project/_api/site/id to get siteId, and https://contoso.sharepoint.com/sites/some-project/_api/web/id to get webId
 
Or running this in PnP Powershell
 
`Grant-PnPAzureADAppSitePermission -AppId 'clientID' -DisplayName 'App Name' -Site 'https://tenant.sharepoint.com/sites/site' -Permissions Write`

The problem was that nobody in the team had permissions to run it, so i escalated it. Microsoft stack is a sea of bureaucracy xd

**Comment by Palanikr** [comment]
@rdinizz Do we need to add azure app registration details(clientid, tenant id) in sharepoint as well. ? to communicate with azure devops pipeline @halliba

**Comment by rdinizz** [comment]
> @rdinizz Do we need to add azure app registration details(clientid, tenant id) in sharepoint as well. ? to communicate with azure devops pipeline @halliba

You need to grant permissions to the app that you created via graph api or pnp powershell. Then it will be able to authenticate and see the site folders from the pipelines.

**Comment by Palanikr** [comment]
@rdinizz @halliba @MHebes 
I have grant permissions in the azure portal by giving access to the following api's

Microsoft GRAPH API Directory.ReadWrite.All Directory.AccessAsUser.All  User.ReadWrite.All Group.ReadWrite.All  GroupMember.ReadWrite.All  Sites.FullControl.All sites.read.all(application) user.read.all
SharePoint Online API AllSites.FullControl  AllSites.Manage  Sites.FullControl.All Sites.Search.All   User.ReadWrite.All   User.ReadWrite.All   user.read.all

Still I'm receiving ""Either scp or roles claim need to be present in the token" error when i run the pipeline
